README.md
# 多功能文档字数统计

多功能文档字数统计，支持中文、日文、朝鲜/韩文，支持选区统计，点击即可高亮，也可用正则表达式自定义要统计的内容。

中国語/日本語/朝鮮/韓国語文字数計算ツール。全体/選択範囲を文字数カウント。強調表示と正規表現(Regex)機能があります。

중국어/일본어/한국어/조선어 세기 도구. 전체/선택 범위를 문자 수 카운트. 강조 및 정규식(Regex) 함수가 있습니다.

Customizable word counter with great support of Chinese characters (Hanzi), Japanese and Korean. Supports count in selection. Supports highlighting. Use custom regexes to match any character type you want.

**[中文介绍](#中文介绍--chinese-description)** / **[English description](#english-description--英文介绍)**

**2023-08-03 更新的 1.6.0 版本为了显示更加简洁，去掉了一些次要的项目，和 UTF-8 显示，如果你认为有必要，可以按[配置教程](./config-guide.md)添加回来**

# 中文介绍 / Chinese Description

## 功能简介

刚安装扩展时，状态栏的右下角会出现一个铅笔的图标，显示字数（需要打开文本文件），鼠标移动到上面会弹出选择语言，请点击你想要显示的语言。下方的工具条也可以随时切换语言。点击「设为默认」可以将对应页面设为默认页面。

在状态栏上显示文档字数，鼠标移动到字数上，会显示详细统计信息。所有内容均可自定义。

点击对应项目的数字可以高亮匹配的文字，在编辑器右边的滚动条上也会标记出匹配文字所在的位置。

该扩展使用正则表达式匹配各类文字，可以在设置中自定义。

状态栏显示默认在右边，也可以在设置里更改。

## 对比其他扩展

**不卡顿，性能好**

统计结果按行缓存，不会因为字数很多就变卡顿。以下是与其他扩展的速度对比([详情](comparison.md#性能对比))。

[<img alt="speed-comparison.png" src="images/speed-comparison.png" width="400px" />](comparison.md#性能对比)

为了便于在图片上表示，将该扩展耗费时间全部归一化。实际上字数多了该扩展花费的时间也会变多一些，但是增长相对其他扩展很小。

**支持多语言：中日韩**

默认配置有英简繁日韩，内容符合对应语言用户所需

<img alt="中文界面" src="images/screenshot-tooltip.png" width="200px" />
<img alt="English" src="images/screenshot-tooltip-english.png" width="200px" />

**结果正确，符合直觉**

正确地统计字数做起来比听起来难。本扩展使用 Unicode 属性决定字符属于哪类；使用现代 JavaScript 分词 API `Intl.Segmenter` 处理组合字和组合符号。对各国文字、emoji 兼容性都极佳！

作者不会日语和(朝鲜语/韩语)，但是这些语言字数统计是准确的。（[详情](comparison.md)）

**点击即可高亮**

无论如何也找不到文档中的某个非 ASCII 字符在哪？只需点一下，就可以把它高亮出来。也是方便的文字类型可视化工具。

<img alt="highlight" src="images/screenshot-highlight.png" width="400px" />

**可自行魔改**

作者「金毛」认为每个工具都应该留下足够大的魔改空间，总有一些深度用户有极强的改造力和控制力。因此，本扩展可以自行添加正则表达式匹配你想要的几乎任何东西，并用 JavaScript 函数模板控制显示内容！([配置教程](config-guide.md))

**可在 Web 版中运行**

本扩展不依赖于 nodejs 或某个特定平台，在浏览器环境下的 [VS Code for the Web](https://vscode.dev/) 中也能运行。

# English Description / 英文介绍

## Basic Introduction

After installing, a pencil icon and character count will appear in the right area of status bar. Pointing at it will reveal a small pop-up window containing a language selection interface. Click the language you want it to show in. The bottom bar of the small window can also be used to change language temporarily. Press `Set as default` to set current page as the default page.

It will show document character count on the status bar. Pointing at it will reveal detailed statistics. All items in it can be customized.

By clicking on numbers, it will highlight respective text, the position of text will also be shown on the scrollbar at right of the text editor.

This extension uses Regular Expression to match any type of character. It can be customized in the configuration.

The status bar item is at right side by default, but you can change it to left in the settings.

## Comparison with other extensions

**No lags, good performance**

The statistics result is calculated and cached per line. It won't lag because of very long text.

This is comparison of time spent to other extensions: ([More info](comparison.md#性能对比))。

[<img alt="speed-comparison.png" src="images/speed-comparison.png" width="400px" />](comparison.md#性能对比)

To make the image prettier, the time spent of this extension is unified. However, the time will increase with more text, but it's much less time compared to other extensions.

**Multiple languages including Chinese, Japanese, and Korean**

The default configuration includes English (Western languages), Simplified and Traditional Chinese, Japanese, and Korean. The items shown is considered as relevant to the language user.

<img alt="English" src="images/screenshot-tooltip-english.png" width="200px" />
<img alt="中文界面" src="images/screenshot-tooltip.png" width="200px" />

**Proper and intuitive result**

It's harder to count words correctly than it seems. This extension uses Unicode properties to decide which type the character belongs to. It also uses modern segmentation JavaScript API `Intl.Segmenter` to process combined characters and symbols. Greatly supports multiple scripts and emoji!

The author hasn't learned Japanese and Korean language, but the results are correct. ([More details](comparison.md))

**Click to highlight**

Can't find out where the non-ASCII character is? Only a click to highlight it. It's also a convenient tool to visualize character types.

<img alt="highlight" src="images/screenshot-highlight.png" width="400px" />

**Hackable by user**

*geezmolycos*, the author, believes that every tool should leave some free space to be modified and hacked. There are always some users who have strong desire to change and controll things around them. Therefore, you can add custom regex to match anything that you want, and change display text with JavaScript function as template! ([Configuration guide](config-guide.md))

**Works in the web version**

The extension doesn't rely on NodeJS or a specific platform. It can run in [VS Code for the Web](https://vscode.dev/).


CHANGELOG.md
# Change Log

All notable changes to the "vscode-hanzi-counter" extension will be documented in this file.

## TODO

- twitter character count
- 日本语较准确的「原稿用紙換算」
- 发布到open vsx上

## [1.6.2] - 2023-10-05

- 不统计过大的文件(>3M字符)或二进制文件，因为会占用过多CPU时间，使extension host假死 (#6)

## [1.6.1] - 2023-08-18

- 修复了西文句数包含空行的问题 (#5)

## [1.6.0] - 2023-08-03

- 给默认状态栏模板中的 `tableRow` 函数添加了可选 `small` 参数，方便将文字变小，节省横向空间。
- 在所有「单词」的正则表达式规则中，排除了类似汉字的文字，因为汉字统计词数规则不同，和西文混合统计不合理 (#3)
- 默认禁用了一些我认为多余的统计条目，让其更简洁
- 将西文页面的「单词」换成分段器的单词，禁用空格分隔的单词，因为这样效果较好
- 将中文页面的「单词数」换成「字词数」，并移动到首位，其计算方式为 CJK 字符（含标点）和西文单词数的和
- 将中文页面的「中文」改为「汉字」
- 默认禁用 UTF-8 详细信息显示，因为对很多人用处不大

## [1.5.2] - 2023-05-08

- 在工具提示里面添加了虚拟的链接，供 VS Code 检测，以使鼠标离开工具提示时不消失 (fix #2)
- 设置了状态栏项目的 `accessibilityInformation` 属性，方便屏幕阅读器阅读

## [1.5.1] - 2023-03-29

- CJK 字符现在使用了分段器

## [1.5.0] - 2023-03-29

- 加入新的正则名字修饰符，表示预先将文本规范化为 Unicode 标准化形式
  - 支持四种规范化形式(nfd/nfc/nfkd/nfkc)，支持高亮
- 谚文匹配规则利用了修饰符，更加简洁
- 更改了模糊的「非空白字符」「非ASCII字符」等，加入码位数，以澄清字符和码位的区别
- 重新排列了各语言的项目顺序，更符合直觉
  - 更新了utf-8项目，版面更加整齐
- 给设置项添加了中文描述
- 设置教程中加入了添加删除项目的教程
- 修改颜色，提供视觉线索
  - 白：主要
  - 红：片假名
  - 橙：平假名
  - 黄：汉字
  - 绿：谚文
  - 蓝：标点
  - 紫：符号

## [1.4.2] - 2023-03-17

- 更新了 `package.json` ，使得扩展可在 VS Code for the Web 中运行

## [1.4.1] - 2023-03-17

- 删除不再使用的 command contribution
- 1.4.0 版本存在下载问题，修复不能直接安装 (fix #1)

## [1.4.0] - 2023-03-11

- 增加了配置教程
- 取消介绍页面，增加 tooltip「设为默认」命令，在模板环境可以获取默认。刚开始会让用户选语言
- README.md 中增加英文
- 更改了各语言的标点符号匹配规则，现在只匹配对应语言文字中的标点符号

## [1.3.3] - 2023-03-10

- 之前打包了多余文件，造成文件变大，现在删除这些多余文件

## [1.3.2] - 2023-03-10

- 将汉字和 CJK 字符包括 Letter_Number 类，使得「〇」和苏州码子也包括在汉字内
- 更改非空白字符规则，添加 Segmenter
- 增加了该扩展和其他工具的正确性对比文档
- 增加了性能对比文档
- 修改了 `README.md` 更直白，易于理解

## [1.3.1] - 2023-03-08

- 使用 Segmenter 的正则表达式也可以正常高亮了
  - 若该正则匹配 segment 标记字符，则会高亮整个 segment
  - 受算法性能限制，限制最大文本长度到 200000 (UTF-16，使用 substring 方法)，极罕见的情况下可能会产生半个surrogate pair
- 数字默认使用逗号分隔，更加清晰

## [1.3.0] - 2023-03-08

- 为正则表达式加入匹配 Grapheme cluster / word / sentence 的选项
  - 使用正则表达式的命名表示需要不需要开启分词器
  - 使用 U+fdd0 和 U+fdd1 表示非单词边界和单词边界
- 更改默认字符数为 grapheme cluster 数，另外提供 codepoint 数

## [1.2.0] - 2023-02-25

- 加入日语、(朝鲜语/韩语)匹配规则
- 加入更多自带的匹配规则，涵盖 unicode 各 General Category，以便使用
- 加入tooltip分页，功能概览页显示设置链接，快速进入设置
  - 功能概览提供简要的说明和快速教程
- 加入日语韩语简介
- 现在找不到模板或者模板代码出错，会弹出错误提示
- 更改高亮的风格，改成粗的下划线

## [1.1.0] - 2023-02-24

- 加入高亮指定正则匹配的指令，可以在模板中加入指令，点击以后在文件中高亮此项的匹配
  - 为了性能考虑，一次最多能高亮20000个区间，如果超过了20000个，再安排20000个到可视区间里
  - 可以一次高亮多种类型的文字，会使用多种颜色区分这些文字。例如可以把中文和中文标点分开高亮
- 给模板代码加入环境this，可以用来记录变量。比如预先生成的查找表就可以缓存到this中。

## [1.0.0] - 2023-02-13

- 为每个打开的文档建立一个DocumentCounter的实例，按行缓存各正则表达式统计的字数，提高性能
  - 只在文件打开、保存时统计全文字数
  - 修改时只统计修改部分的字数，并更新全文字数
  - 在编辑字数很多的文件时，性能有显著提升
- 修复了建立多个选区时字数统计只考虑第一个选区的问题
- 增加启用禁用设置项，可以指定仅某语言下启用
  - 例如可以仅在编辑markdown时显示统计字数
- tooltip的模板可以添加任意个，使用指令链接切换
  - 例如分别设置中文统计、日文统计、英文统计的tooltip模板，显示对应的内容
- 分离了源码文件，更改了vscode导入方式，更加方便维护
- MIT License

## [0.1.1] - 2023-02-09

- 缩减中文标点符号范围，缩减到Script_extensions为Han或Common，且分类为P、S、N的字符
- 增加启用禁用、调节状态栏显示的设置

## [0.1.0] - 2023-02-05

- 在状态栏上显示文档字数，tooltip 中显示详细信息，点击展开显示更详细信息
- 在设置中添加要统计字数的各类项目对应的正则表达式
- 在设置中通过 JavaScript 函数自定义要显示的内容
- 默认设置添加统计项目：
  - 单词数
  - 总字符数
  - 非 ASCII 字符数
  - 中文字数（含标点和不含标点）
  - UTF-8 大小


src/counter.ts

import * as vscode from 'vscode';
import { DocumentCounter } from './documentCounter';

const MAX_HIGHLIGHT_COUNT = 20000;
const MAX_HIGHLIGHT_TEXT_LENGTH = 200000;
const MAX_HIGHLIGHT_COUNT_VISIBLE = 20000;

type NormalizationForm = 'nfd' | 'nfc' | 'nfkd' | 'nfkc';

function compileTemplateFunction(parameters: string[], template: string): Function {
    let statementMatchResult = template.match(/^\s*{(.*)}\s*$/);
    let expressionMatchResult = template.match(/^\s*(.*)\s*$/);
    let functionBody: string;
    if (statementMatchResult){
        functionBody = statementMatchResult[1];
    } else if (expressionMatchResult){
        functionBody = 'return ' + expressionMatchResult[1];
    } else {
        throw new Error('template is not a valid expression or code block');
    }
    return new Function(...parameters, functionBody);
}

export class Counter {

    public readonly regexes: Map<string, RegExp>;
    public readonly segmenters: Map<string, Intl.Segmenter | undefined>; // Intl.segmenter configurations
    public readonly normalizations: Map<string, NormalizationForm | undefined>; // regex norm modifier

    public readonly templateParameters: string[]; // parameter names list for template functions
    public readonly templates: Map<string, Function>;
    public readonly templateEnvironment: {[key: string]: any};

    private _statusBarItem: vscode.StatusBarItem;
    private _decorationTypes: vscode.TextEditorDecorationType[];

    constructor(configuration: vscode.WorkspaceConfiguration) {
        // read and parse configurations
        const regexStrings = new Map(Object.entries(
            configuration.get('vscode-hanzi-counter.counter.regexes') as object
        ));
        this.regexes = new Map();
        this.normalizations = new Map();
        this.segmenters = new Map();
        this.templateParameters = [];
        for (let [k, v] of regexStrings){
            let result = k.split('@');
            if (this.regexes.has(result[0])){
                throw new Error(`regex table contains same keys "${result[0]}"`)
            }

            // set regex map item
            this.regexes.set(result[0], new RegExp(v, 'gus'));
            this.segmenters.set(result[0], undefined);
            this.normalizations.set(result[0], undefined);
            this.templateParameters.push(result[0]);
            // has segment
            if (result.length >= 2 && result[1].length > 0){
                let granularity = new Map<string, Intl.SegmenterOptions['granularity']>([
                    ['g', 'grapheme'],
                    ['w', 'word'],
                    ['s', 'sentence']
                ]).get(result[1][0]);
                if (granularity === undefined){
                    throw new Error(`invalid granularity code for segment "${result[1][0]}" in regex "${k}", it must be one of "gws"`);
                }
                let locale = result[1].substring(1) || undefined;
                this.segmenters.set(result[0], new Intl.Segmenter(locale, {"granularity": granularity}));
            }
            // has normalization form modifier
            if (result.length >= 3 && result[2].length > 0){
                result[2] = result[2].toLowerCase();
                if (result[2] !== 'nfc' && result[2] !== 'nfd' && result[2] !== 'nfkd' && result[2] !== 'nfkc'){
                    throw new Error(`invalid normalization form "${result[2]}"`);
                }
                let normForm: NormalizationForm = result[2] as NormalizationForm;
                this.normalizations.set(result[0], normForm);
            }
        }

        const templateStrings = new Map(Object.entries(
            configuration.get('vscode-hanzi-counter.counter.templates') as object
        ));
        this.templates = new Map();
        for (let [k, v] of templateStrings){
            this.templates.set(k, compileTemplateFunction(this.templateParameters, v));
        }

        // environment for template function to store variables
        this.templateEnvironment = {'regexes': this.regexes, 'templatesParameters': this.templateParameters, 'templates': this.templates, 'segmenters': this.segmenters};

        // create status bar item
        this._statusBarItem = vscode.window.createStatusBarItem(
            configuration.get('vscode-hanzi-counter.statusBar.alignment') === 'left'
                ? vscode.StatusBarAlignment.Left : vscode.StatusBarAlignment.Right,
            configuration.get('vscode-hanzi-counter.statusBar.priority') ?? 105); // default left of text attributes(ln, col, spaces, encoding, etc)
        this._statusBarItem.name = 'Hanzi Counter';
        this._statusBarItem.accessibilityInformation = {label: 'Hanzi Counter'};

        // create decoration type
        this._decorationTypes = [
            ['', 'editor.findMatchHighlightBackground', 'charts.foreground', 'charts.foreground'],
            ['', 'editor.findMatchHighlightBackground', 'charts.red', 'charts.red'],
            ['', 'editor.findMatchHighlightBackground', 'charts.orange', 'charts.orange'],
            ['', 'editor.findMatchHighlightBackground', 'charts.yellow', 'charts.yellow'],
            ['', 'editor.findMatchHighlightBackground', 'charts.green', 'charts.green'],
            ['', 'editor.findMatchHighlightBackground', 'charts.blue', 'charts.blue'],
            ['', 'editor.findMatchHighlightBackground', 'charts.purple', 'charts.purple'],
        ].map(([color, background, border, ruler]) => vscode.window.createTextEditorDecorationType({
            'color': color ? new vscode.ThemeColor(color) : undefined,
            'backgroundColor': new vscode.ThemeColor(background),
            'borderColor': new vscode.ThemeColor(border),
            'borderWidth': '0 0 0.5em 0',
            'borderStyle': 'solid',
            'overviewRulerColor': new vscode.ThemeColor(ruler),
            // should not extend
            'rangeBehavior': vscode.DecorationRangeBehavior.ClosedClosed
        }));
    }

    public changeStatusBarItem(show: boolean){
        if (show){
            this._statusBarItem.show();
        } else {
            this._statusBarItem.hide();
        }
    }

    public updateStatusBarItem(text?: string, tooltipText?: string){
        if (text){
            this._statusBarItem.text = text;
            this._statusBarItem.accessibilityInformation = {label: text};
        }
        if (tooltipText){
            let ms = new vscode.MarkdownString('<!--[Fake link](#)-->\n' + tooltipText);
            // The "Fake link" part is to let vscode think it contains a markdown link
            // so that the tooltip pop-up will stay open on mouse leave
            // Reference: https://github.com/microsoft/vscode/blob/d1aa00acef8b2cfa88621c448b8fd8cd034f60a9/src/vs/workbench/services/hover/browser/hoverWidget.ts#L205
            ms.isTrusted = true;
            ms.supportHtml = true;
            ms.supportThemeIcons = true;
            this._statusBarItem.tooltip = ms;
        }
    }

    public setHighlight(ranges: readonly vscode.Range[], typeNumber: number){
        vscode.window.activeTextEditor?.setDecorations(this._decorationTypes[typeNumber], ranges);
    }

    public removeHighlight(){
        for (let i = 0; i < this._decorationTypes.length; ++i){
            this.setHighlight([], i);
        }
    }

    public setHighlightRegex(regexNames: string | string[] | string[][]){
        this.removeHighlight();
        let regexGroupList: string[][];
        if (typeof regexNames === 'string'){
            regexGroupList = [[regexNames]];
        } else if (regexNames.length === 0){
            return;
        } else if (typeof regexNames[0] === 'string'){
            regexGroupList = regexNames.map(_ => [_]) as string[][];
        } else {
            regexGroupList = regexNames as string[][];
        }
        let regexGroups = regexGroupList.map(rList => rList.map(
            r => [this.regexes.get(r), this.segmenters.get(r), this.normalizations.get(r)] as
            [RegExp | undefined, Intl.Segmenter | undefined, NormalizationForm | undefined]
        ));
        let currentDocument = vscode.window.activeTextEditor?.document;
        if (currentDocument){
            let eolString: string = new Map([
                [vscode.EndOfLine.LF, '\n'], [vscode.EndOfLine.CRLF, '\r\n']
            ]).get(currentDocument.eol)!;
            let selections = vscode.window.activeTextEditor!.selections;
            let allEmpty = true;
            for (let selection of selections){ // handle multi-selection
                if (!selection.isEmpty){
                    allEmpty = false;
                    break;
                }
            }
            let selectionRanges: vscode.Range[] = selections.slice();
            if (allEmpty){ // no text is selected
                selectionRanges = [new vscode.Range(0, 0, currentDocument.lineCount, 0)]; // entire document
            }
            for (let i = 0; i < regexGroups.length; ++i){
                let regexes = regexGroups[i];
                if (!regexes){
                    continue;
                }
                let highlightRanges: vscode.Range[] = [];
                for (let regexTuple of regexes){
                    if (!regexTuple[0]){
                        continue;
                    }
                    let addSelectionRangeHighlight = (selectionRange: vscode.Range, maxCount: number) => {
                        let [regex, segmenter, norm] = regexTuple;
                        let normalization = norm?.toUpperCase();
                        let startOffset = currentDocument!.offsetAt(selectionRange.start);
                        let text = currentDocument!.getText(selectionRange);
                        if (text.length > MAX_HIGHLIGHT_TEXT_LENGTH){
                            text = text.substring(0, MAX_HIGHLIGHT_TEXT_LENGTH);
                        }

                        // normalization form processing
                        // calculate string position lookup, similar to segmenter
                        let getBeforeNormalizationOffset: (offset: number) => number;

                        if (normalization === undefined){
                            getBeforeNormalizationOffset = (offset) => offset;
                        } else if (normalization === 'nfd' || normalization === 'nfkd') {
                            let beforeText = text;
                            let beforeTextIterator = beforeText[Symbol.iterator]();
                            let afterText = text.normalize(normalization);
                            text = afterText;
                            let beforeOffset = 0;
                            let afterOffset = 0;
                            getBeforeNormalizationOffset = (offset) => { // works for nfd and nfkd
                                while (offset > afterOffset){
                                    let bef = beforeTextIterator.next().value; // handles utf-16 surrogate pair
                                    let aft = bef.normalize(normalization);
                                    beforeOffset += bef.length;
                                    afterOffset += aft.length;
                                }
                                return beforeOffset;
                            };
                        } else {
                            let beforeText = text;
                            let beforeTextIterator = beforeText[Symbol.iterator]();
                            let dNormalization = normalization.replace('C', 'D');
                            let dText = text.normalize(dNormalization);
                            let cText = dText.normalize(normalization);
                            text = cText;
                            let beforeOffset = 0;
                            let dOffset = 0;
                            let dcOffset = 0;
                            let cOffset = 0;
                            getBeforeNormalizationOffset = (offset) => {
                                // this is somewhat difficult to explain
                                // if you need to understand this piece of code
                                // wish you good luck
                                let cNew = cText.substring(cOffset, offset);
                                let dcNew = cNew.normalize(dNormalization);
                                cOffset += cNew.length;
                                dcOffset += dcNew.length;
                                while (dcOffset > dOffset){
                                    let bef = beforeTextIterator.next().value;
                                    let d = bef.normalize(dNormalization);
                                    beforeOffset += bef.length;
                                    dOffset += d.length;
                                }
                                return beforeOffset;
                            };
                        }

                        let getBeforeSegmentingOffset: (offset: number) => number, getSegmentEndOffset: (offset: number) => number;
                        // segment text
                        if (segmenter !== undefined){
                            let textLines = text.split(eolString);
                            let segmentText = '';
                            // split to lines for faster segmenting
                            // Intl.Segmenter is slow for large input
                            for (let line of textLines){
                                let lineSegmentText = DocumentCounter.addSegmentIndicators(line, segmenter);
                                segmentText += lineSegmentText + eolString;
                            }
                            text = segmentText.substring(0, segmentText.length - eolString.length); // remove final eol string
                            let indicatorMatches = Array.from(text.matchAll(/[\ufdd0-\ufdef]/g));
                            let sentinel: RegExpMatchArray = ['\ufdd0'];
                            sentinel.index = text.length;
                            indicatorMatches.push(sentinel);
                            let nextIndicatorOffset = indicatorMatches[0].index ?? Infinity;
                            let indicatorCount = 0;
                            // offset must not decrease for correct result
                            getBeforeSegmentingOffset = (offset: number) => {
                                if (offset === nextIndicatorOffset){
                                    return offset - indicatorCount;
                                }
                                if (offset < nextIndicatorOffset){
                                    return offset - indicatorCount;
                                }
                                while (offset > nextIndicatorOffset){
                                    indicatorCount += 1;
                                    nextIndicatorOffset = indicatorMatches[indicatorCount].index ?? Infinity;
                                }
                                return getBeforeSegmentingOffset(offset);
                            };
                            // get segment end if character at offset is segment indicator
                            getSegmentEndOffset = (offset: number) => {
                                let maybeIndicatorOffset = getBeforeSegmentingOffset(offset - 1);
                                if (offset - 1 === nextIndicatorOffset){ // match ends with indicator
                                    return (indicatorMatches[indicatorCount + 1].index ?? Infinity) - indicatorCount - 1;
                                }
                                return getBeforeSegmentingOffset(offset);
                            };
                        } else {
                            // no segmenter, range is as before
                            getBeforeSegmentingOffset = getSegmentEndOffset = (offset: number) => {
                                return offset;
                            };
                        }
                        for (let match of text.matchAll(regex!)){
                            let matchStartIndex = getBeforeNormalizationOffset(getBeforeSegmentingOffset(match.index!));
                            let matchEndIndex = getBeforeNormalizationOffset(getSegmentEndOffset(match.index! + match[0].length));
                            let matchRange = new vscode.Range(
                                currentDocument!.positionAt(startOffset + matchStartIndex),
                                currentDocument!.positionAt(startOffset + matchEndIndex)    
                            );
                            highlightRanges.push(matchRange);
                            if (highlightRanges.length >= maxCount){
                                // too many highlight ranges, give up
                                return false;
                            }
                        }
                        return true;
                    };
                    for (let selectionRange of selectionRanges){
                        addSelectionRangeHighlight(selectionRange, MAX_HIGHLIGHT_COUNT);
                    }
                    // consider visible regions
                    if (highlightRanges.length >= MAX_HIGHLIGHT_COUNT){
                        for (let visibleRange of vscode.window.activeTextEditor!.visibleRanges){
                            let hasSpaceLeft = addSelectionRangeHighlight(visibleRange, MAX_HIGHLIGHT_COUNT + MAX_HIGHLIGHT_COUNT_VISIBLE);
                            if (!hasSpaceLeft){
                                break;
                            }
                        }
                    }
                }
                this.setHighlight(highlightRanges, i);
            }
        }
    }

    public dispose() {
        this._statusBarItem.dispose();
        for (let t of this._decorationTypes){
            t.dispose();
        }
    }
}
